### ISSUE TITLE
DownsamplerBlock torch cat inconsistent tensor size

### ISSUE THREAD

**Issue by zhhtu** [Eromera/erfnet_pytorch#8]
hi Eromera, I'm using my dataset to train the model, but it seems something wrong with the torch.cat operation in DownsamplerBlock:

![screen shot 2017-12-14 at 8 51 21 am](https://user-images.githubusercontent.com/19236599/33970089-0626f8dc-e0ac-11e7-951b-b3d2f2af91b2.png)

it seems that results from conv and pool doesnot match for odd image size...
Also, due to the repeated usage of this "DownsamplerBlock", all "cat" operation input must have the even size. 
Also, I'm curious about this "cat" operation, does it contribute to the final performance? I mean why just use the conv or pool operation directly?

**Comment by Eromera** [comment]
Hi, what is the resolution you are using for your dataset? Since there are 3 downsampling operations in the encoder, each one dividing size by 2, the input has to be divisible by 8. Otherwise, Pytorch will throw that kind of error (inconsistent tensor sizes) at the end of the downsampling block (the concatenation operation). If the resolution is not divisible by 8, you can either resize it or crop the last pixels that exceed width and height (you can use module operation %8 to check exceeding values)

The torch.cat operation is concatenating the output feature maps of the maxpool layer and the conv layer in the downsampler block. If you don't want to use it you can remove it and use only max pooling layers or only conv layers with stride=2 to do the downsampling operation, but since this would hurt accuracy and since your problem is not related to the cat operation then I would't advise that.

**Comment by zhhtu** [comment]
Thank you for your replay Eromera, I've cut off some pixels for my own dataset. But another error pumps up:
![screen shot 2017-12-14 at 1 50 14 pm](https://user-images.githubusercontent.com/19236599/33977500-1cf2b2de-e0d6-11e7-9c6a-8cb8c232c1aa.png)

I'm not quite sure where the problem is, I checked the label file there's no "-1" label and I've changed the default igoreIndex in the iouEval file, also my background class is the last one among the labels

**Comment by Eromera** [comment]
That error should be related to labels that are not in the ground truth, not only -1 but also any other value that you could have introduced when doing crops or some other operation. You can debug the images before doing forward by printing numpy.unique(label.numpy()) where label is the torch tensor. It will print the list of unique values that are available and you should be able to see if there are any weird values. For example, for cityscapes it should be a value between 0-19. Otherwise there is something wrong with the preprocessing. Also, are you cropping the same regions in both images and labels? (they must be same size)

**Comment by zhhtu** [comment]
hi Eromera, I've conformed the error. It is the label that caused the problem, thanks very much.
